### PR TITLE
Adding defaultReject feature and _instanceId property (resolves #3984)

### LINKED ISSUE #3984: Need a way to handle rejection/resolve after request is completed
#### Is your feature request related to a problem? Please describe.
A problem:
Currently axios contains interceptors, which are executed before standard handlers.
What we want to do: we want to have standard error handler, which will execute only if there is no local handler

#### Describe the solution you'd like
```javascript
import Toast from 'toast';

const axiosInstance = axios.create();

axiosInstance.defaultReject((error) => {
  if (['401', '403', '404'].includes(error.status.code)) {
    Toast(error);
  }
  // in case we don't expect that issue at all we fail-fast
  return Promise.reject();
})

// guess also makes sense for some cases
axiosInstance.defaultResolve();

axiosinstance.get('whatever').catch((error) => {
  if (error.status.code === 404) {
    // show user that something was not found
  }
  // axiosInstance.defaultReject will execute
  return Promise.reject(error);
})

// in case of error axiosInstance.defaultReject will execute
axiosinstance.get('whatever-another');

axiosinstance.get('whatever-another-smoething).catch((error) => {
  // only that handler will execute, and not axiosInstance.defaultReject
});
```

#### Describe alternatives you've considered
Currently only 1 alternative was found - copypaste "standardhandler" everywhere in catch

#### Additional context
no

### PR DESCRIPTION
**Pull request by theo-js**
The only way I could find to implement this feature was by listening to the 'unhandledRejection' event.
I've added an _instanceId property to axios instances, that is incremented every time a new instance is created; this way, when the unhandledRejection event happens, only the defaultReject of the same instance that sent the request is executed.

However, I was unable to come up with a way to test this, as Jasmine does not accept unhandled promise rejections ...

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -478,6 +478,13 @@ These are the available config options for making requests. Only the `url` is re
     
     // throw ETIMEDOUT error instead of generic ECONNABORTED on request timeouts
     clarifyTimeoutError: false,
+  },
+
+  // `defaultReject`, when set to a function, is executed when your axios request is rejected
+  // and you haven't explicitly used a catch block
+  // it implements the unhandledRejection event
+  defaultReject: function (err) {
+    console.log(err.toJSON());
```
Comment (Hulkmaster) [inline README.md:487]: maybe `console.error`?

also not sure about axios guidelines, but i would use full names
https://javascript.info/ninja-code#use-abbreviations

Comment (Hulkmaster) [reply README.md:487]: also i think in order to not to break backward compatibility it makes sense to `throw error` here by default

File: README.md [hunk 2]
```diff
@@ -478,6 +478,13 @@ These are the available config options for making requests. Only the `url` is re
     
     // throw ETIMEDOUT error instead of generic ECONNABORTED on request timeouts
     clarifyTimeoutError: false,
+  },
+
+  // `defaultReject`, when set to a function, is executed when your axios request is rejected
+  // and you haven't explicitly used a catch block
+  // it implements the unhandledRejection event
+  defaultReject: function (err) {
```
Comment (Hulkmaster) [inline README.md:486]: would also add `defaultResolve`

File: lib/axios.js [hunk 3]
```diff
@@ -27,6 +28,35 @@ function createInstance(defaultConfig) {
...
+      promise.catch(function catchUnhandledPromise() {
+        instance.defaults.defaultReject(axiosError);
+        if (typeof unhandledPromiseEvent.preventDefault === 'function') {
+          unhandledPromiseEvent.preventDefault();
+        }
+      });
+    }
+  }
+
+  if (typeof process !== 'undefined' && Object.prototype.toString.call(process) === '[object process]') {
```
Comment (Hulkmaster) [inline lib/axios.js:48]: maybe create a helper function `isNodeEnvironment`?

File: lib/axios.js [hunk 4]
```diff
@@ -27,6 +28,35 @@ function createInstance(defaultConfig) {
...
+    // For node use process.addListener
+    process.addListener('unhandledRejection', function handleServerSideUnhandledRejection(e, promise) {
+      useDefaultReject(e, e, promise);
+    });
+  } else if (typeof window !== 'undefined') {
+    // For browsers use window.addEventListener
+    window.addEventListener('unhandledrejection', function handleBrowserSideUnhandledRejection(e) {
```
Comment (Hulkmaster) [inline lib/axios.js:55]: not sure why you use here named function

Comment (theo-js) [reply lib/axios.js:55]: Running tests forces me to do so

File: README.md (hunk 2 above)
Comment (theo-js) [reply README.md:486]: I don't know how to listen to unhandled promise resolutions

File: README.md (hunk 1 above)
Comment (theo-js) [reply README.md:487]: Makes sense ! So far, by default, defaultReject is null

**Comment by Hulkmaster** [comment]
btw, i'm don't know axios code, but i think there might be easier way to do it

axios returns a promise, so somewhere in the code there should be something like

```
return new Promise((resolve, reject) => {
  const request = new XMLHttpRequest();
  ...
  resolve(request.result);
})
```

can be modified to

```
const promise = new Promise(...);

if (settings.defaultResolve) {
  promise.then(settings.defaultResolve);
}

if (settings.defaultReject) {
  promise.catch(settings.defaultReject);
}
if (settings.defaultFinally) {
  promise.finally(settings.defaultFinally);
}

return promise;
```

because listening to unhandledRejection sounds like something, which can theoretically bring side-effects 🤔

**Comment by theo-js** [comment]
I agree, but the problem with your solution is that settings.defaultReject will always be called, no matter if you do a .catch() after axios.request(), or not;

If we want the defaultReject handler to be called for axios.request(), but not for axios.request().catch(), this cannot work, because at that stage, the promise has no way to know if it will later be caught

**Comment by Hulkmaster** [comment]
ah, true

but then i don't think its a big issue actually. i would then just call it `afterResolve` `afterReject` and `afterFinally`

and then consumer will have to deal with "trying to understand if is unhandled before", don't think its a big issue, and still provides with usefull API, with minimum risks

**Comment by Hulkmaster** [comment]
> Isn't that what axios interceptors already do solve ?
yes, my bad //_-

**Comment by Hulkmaster** [comment]
Okay, got crazier idea

So promise, basically, has an array of `then` and `catch` callbacks, which it executes one by one, so what we want is kinda put "one final callback" in the end of each of each queue, which i had not found a way to do

But!

We can extend existing Promise with new features, like `defaultResolve` and `defaultReject`

so 
```AxiosPromise extend Promise {
  defaultResolve() {}
  defaultReject() {}
}```

**Comment by jasonsaayman** [comment]
I think should a user want to implement something like this I think inteceptors should be used.

**Pull request closed without merging**